src/inference_engines/mlc_engine.py

```diff
@@ -0,0 +1,188 @@
+import asyncio
+import json
+import os
+import time
+from io import BytesIO, IOBase
+from os import path
+from queue import Queue
+from threading import Lock, Thread
+from typing import BinaryIO, List, Optional, Union, get_args
+
+import torch
+from cog import ConcatenateIterator
+from mlc_chat import ChatConfig, ChatModule, ConvConfig, GenerationConfig
+from mlc_chat.callback import StreamIterator
+from src.config_utils import Weights
+from transformers import AutoTokenizer
+
+from .engine import Engine
+
+
+class AsyncCompletionStream:
+    def __init__(self, cm: ChatModule, generation_config: GenerationConfig):
+        self.generation_config = generation_config
+        self.cm = cm
+
+    def __aiter__(self):
+        return self
+
+    async def get_next_msg(self):
+        if not self.cm._stopped():
+            self.cm._decode(generation_config=self.generation_config)
+            msg = self.cm._get_message()
+            return msg
+        else:
+            raise StopAsyncIteration
+
+    async def __anext__(self):
+        if not self.cm._stopped():
+            task = asyncio.create_task(self.get_next_msg())
+            msg = await task
+            return msg
+        else:
+            raise StopAsyncIteration
```

Comment:
I don't think this makes sense. aiter and anext aren't used at all. get_next_msg doesn't benefit from being async in any way; it doesn't use await, so it doesn't yield control. _decode or _get_message would need to be async for that to work.

Reply:
resolved in a different PR